src/cascadia/TerminalApp/Jumplist.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "pch.h"
+#include "Jumplist.h"
+
+#include <Propkey.h>
+
+using namespace winrt::TerminalApp;
+
+//  This property key isn't already defined in propkey.h, but is used by UWP Jumplist to determine the icon of the jumplist item.
+//  IShellLink's SetIconLocation isn't going to read "ms-appx://" icon paths, so we'll need to use this to set the icon.
+DEFINE_PROPERTYKEY(PKEY_AppUserModel_DestListLogoUri, 0x9F4C2855, 0x9F79, 0x4B39, 0xA8, 0xD0, 0xE1, 0xD4, 0x2D, 0xE1, 0xD5, 0xF3, 29);
+#define INIT_PKEY_AppUserModel_DestListLogoUri                                             \
+    {                                                                                      \
+        { 0x9F4C2855, 0x9F79, 0x4B39, 0xA8, 0xD0, 0xE1, 0xD4, 0x2D, 0xE1, 0xD5, 0xF3 }, 29 \
+    }
+
+// Method Description:
+// - Updates the items of the Jumplist based on the given settings.
+// Arguments:
+// - settings - The settings object to update the jumplist with.
+// Return Value:
+// - <none>
+void Jumplist::UpdateJumplist(std::shared_ptr<::TerminalApp::CascadiaSettings> settings)
+{
+    winrt::com_ptr<ICustomDestinationList> jumplistInstance;
+    if (FAILED(CoCreateInstance(CLSID_DestinationList, nullptr, CLSCTX_ALL, IID_PPV_ARGS(&jumplistInstance))))
+    {
+        return;
+    }
+
+    // Start the Jumplist edit transaction
+    uint32_t slots;
+    winrt::com_ptr<IObjectCollection> jumplistItems;
+    if (FAILED(jumplistInstance->BeginList(&slots, IID_PPV_ARGS(&jumplistItems))))
+    {
+        return;
+    }
+
+    // It's easier to clear the list and re-add everything. The settings aren't
+    // updated often, and there likely isn't a huge amount of items to add.
+    jumplistItems->Clear();
+
+    // Update the list of profiles.
+    if (FAILED(_updateProfiles(jumplistItems, settings->GetProfiles())))
+    {
+        return;
+    }
+
+    // TODO: Add items from the future customizable new tab dropdown as well.
+    // This could either replace the default profiles, or be added alongside them.
+
+    // Add the items to the jumplist Task section.
+    // The Tasks section is immutable by the user, unlike the destinations
+    // section that can have its items pinned and removed.
+    jumplistInstance->AddUserTasks(jumplistItems.get());
+
+    if (FAILED(jumplistInstance->CommitList()))
+    {
+        return;
+    }
+}
+
+// Method Description:
+// - Creates and adds a ShellLink object to the Jumplist for each profile.
+// Arguments:
+// - jumplistItems - The jumplist item list
+// - profiles - The profiles to add to the jumplist
+// Return Value:
+// - S_OK or HRESULT failure code.
+HRESULT Jumplist::_updateProfiles(winrt::com_ptr<IObjectCollection>& jumplistItems, const gsl::span<const Profile>& profiles)

```

Comment:
when passing COM objects around as parameters, prefer `IComType*`. This ensures that the caller can use whatever container type it wants.

Reply:
(the way it is, it _requires_ the caller to use `winrt::com_ptr` which means we'll need to change the signature if we move to winrt::com_ptr or what-have-you

Reply:
)